channeldb/messagestore.go

```diff
@@ -0,0 +1,204 @@
+		_, err := tx.CreateBucketIfNotExists(messageStoreBucket)
+		return err
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create required buckets: %v",
+			err)
+	}
+
+	return &GossipMessageStore{db}, nil
+}
+
+// oldMessageStoreKey constructs the old message store key for
+// AnnounceSignatures messages. This key will be used first when attempting to
+// retrieve a AnnounceSignatures message from the store in order to provide
+// backwards-compatibility and avoid a database migration.
+func oldMessageStoreKey(msg *lnwire.AnnounceSignatures, peerPubKey [33]byte) []byte {
+	var k [33 + 8]byte
+	copy(k[:33], peerPubKey[:])
+	byteOrder.PutUint64(k[33:41], msg.ShortChannelID.ToUint64())
+
+	return k[:]
+}
+
+// messageStoreKey constructs the database key for the message to be stored.
+func messageStoreKey(msg lnwire.Message, peerPubKey [33]byte) ([]byte, error) {
+	var shortChanID lnwire.ShortChannelID
+	switch msg := msg.(type) {
+	case *lnwire.AnnounceSignatures:
+		shortChanID = msg.ShortChannelID
+	case *lnwire.ChannelUpdate:
+		shortChanID = msg.ShortChannelID
+	default:
+		return nil, ErrUnsupportedMessage
+	}
+
+	var k [33 + 8 + 2]byte
+	copy(k[:33], peerPubKey[:])
+	byteOrder.PutUint64(k[33:41], shortChanID.ToUint64())
+	byteOrder.PutUint16(k[41:43], uint16(msg.MsgType()))
+
+	return k[:], nil
+}
+
+// AddMessage adds a message to the store for this peer.
+func (s *GossipMessageStore) AddMessage(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// Construct the key for which we'll find this message with in the store.
+	msgKey, err := messageStoreKey(msg, peerPubKey)
+	if err != nil {
+		return err
+	}
+
+	// Serialize the message with its wire encoding.
+	var b bytes.Buffer
+	if _, err := lnwire.WriteMessage(&b, msg, 0); err != nil {
+		return err
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		messageStore := tx.Bucket(messageStoreBucket)
+		if messageStore == nil {
+			return ErrCorruptedMessageStore
+		}
+
+		return messageStore.Put(msgKey, b.Bytes())
+	})
+}
+
+// DeleteMessage deletes a message from the store for this peer.
+func (s *GossipMessageStore) DeleteMessage(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// Construct the key for which we'll find this message with in the
+	// store.
+	msgKey, err := messageStoreKey(msg, peerPubKey)
+	if err != nil {
+		return err
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		messageStore := tx.Bucket(messageStoreBucket)
+		if messageStore == nil {
+			return ErrCorruptedMessageStore
+		}
+
+		// If we're attempting to delete an AnnounceSignatures message
+		// from the store, we'll first check if it's using the legacy
+		// key format. We'll only proceed to delete it if it exists,
+		// otherwise, we'll go ahead with the new key format.
+		if msg, ok := msg.(*lnwire.AnnounceSignatures); ok {
+			oldKey := oldMessageStoreKey(msg, peerPubKey)
+			if messageStore.Get(oldKey) != nil {
+				return messageStore.Delete(oldKey)
```

Comment:
Could we risk both keys being in the store? Seems like we might as well just delete both keys always.